src/main/kotlin/com/github/vol0n/utbotcppclion/services/Client.kt

```diff
@@ -0,0 +1,276 @@
+import com.github.vol0n.utbotcppclion.ui.UTBotRequestProgressIndicator
+import com.github.vol0n.utbotcppclion.utils.createFileAndMakeDirs
+
+import testsgen.Testgen
+import testsgen.Util
+
+import com.intellij.openapi.diagnostic.Logger
+import com.intellij.openapi.project.Project
+import kotlin.random.Random
+
+import kotlinx.coroutines.CoroutineExceptionHandler
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.catch
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.swing.Swing
+import kotlinx.coroutines.withContext
+import kotlinx.coroutines.withTimeout
+
+class Client(val project: Project) {
+    private var stub = GrpcStarter.startClient().stub
+    val grpcCoroutineScope: CoroutineScope
+    private var connectionStatus = ConnectionStatus.INIT
+    private val connectionChangedPublisher =
+        project.messageBus.syncPublisher(UTBotConnectionChangedNotifier.CONNECTION_CHANGED_TOPIC)
+    private var heartBeatJob: Job? = null
+    private val logger = Logger.getInstance(this::class.java)
+    private val metadata: io.grpc.Metadata = io.grpc.Metadata()
+
+    init {
+        val handler = CoroutineExceptionHandler { _, exception ->
+            exception.printStackTrace()
+        }
+        grpcCoroutineScope = CoroutineScope(Dispatchers.Swing + handler)
+
+        periodicHeartBeat()
+    }
+
+    fun getConnectionStatus() = connectionStatus
+    fun isServerAvailable() = connectionStatus == ConnectionStatus.CONNECTED
+
+    fun doHandShake() {
+        grpcCoroutineScope.launch {
+            try {
+                stub.handshake(Testgen.DummyRequest.newBuilder().build())
+            } catch (e: Exception) {
+                logger.warn("HandShake failed with the following error: ${e.message}")
+            }
+        }
+    }
+
+    private suspend fun setMetadata() {
+        fun createRandomSequence() = (1..RANDOM_SEQUENCE_LENGTH)
+                .map { Random.nextInt(0, RANDOM_SEQUENCE_MAX_VALUE).toString() }
+
+        val clientID = project.name + (System.getenv("USER") ?: "unknownUser") + createRandomSequence()
+        try {
+            stub.registerClient(Testgen.RegisterClientRequest.newBuilder().setClientId(clientID).build())
+        } catch (e: Exception) {
+            logger.warn("Setting metadata failed: ${e.message}")
+        }
+        metadata.put(io.grpc.Metadata.Key.of("clientid", io.grpc.Metadata.ASCII_STRING_MARSHALLER), clientID)
+        stub = io.grpc.stub.MetadataUtils.attachHeaders(stub, metadata)
+    }
+
+    fun periodicHeartBeat() {
+        if (heartBeatJob != null) {
+            heartBeatJob?.cancel()
+        }
+        logger.info("Started heartbeating the server!")
+        heartBeatJob = grpcCoroutineScope.launch {
+            setMetadata()
+            while (true) {
```

Comment:
Please add comment about https://plugins.jetbrains.com/docs/intellij/plugin-services.html that it is ok

Reply:
I changed it in the last commit: added a dispose method, that cancels all the launched coroutines